compiler/rustc_target/src/target_features.rs

```diff
@@ -1177,6 +1178,14 @@ impl Target {
                     _ => unreachable!(),
                 }
             }
+            "s390x" => {
+                // We don't currently support a softfloat target on this architecture.
+                // As usual, we have to reject swapping the `soft-float` target feature.
+                // The "vector" target feature does cause floats to be put into vector
+                // registers, but those overlap the float registers so it makes no
+                // difference for the ABI.
```

Comment:
Based on what @uweigand said, the "vector" feature's effects on float handling is entirely intra-function as opposed to cross-function and thus isn't ABI-relevant? Perhaps it should instead be
```suggestion
                // The "vector" target feature does not affect the ABI for floats,
                // and vector registers overlap the float registers anyways.
```

Reply:
I believe, based on what @uweigand just said, that there's 16 *extra* vector registers with the "vector" feature, and those can _also_ be used to hold floats. That is... let's give ourselves 32 logical registers:

| r0 | r1 | r2 | r3 | r4 | r5 | r6 | r7 | r8 | r9 | r10 | r11 | r12 | r13 | r14 | r15 | r16 | r17 | r18 | r19 | r20 | r21 | r22 | r23 | r24 | r25 | r26 | r27 | r28 | r29 | r30 | r31 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| FP | FP | FP | FP | FP | FP | FP | FP | FP | FP | FP | FP | FP | FP | FP | FP | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - |
| VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR | VR |

So for instructions that are designed to address the old floating-point range, they likely have 4 bits for addressing in their opcode and still only work on the "bottom halves" of the vector registers that occupy r0 through r15. But new vector operations, including "upgraded" versions of old float instructions, can still do float operations, and thus can manipulate r16 through r31 as if they are floating point registers.

However, it still breaks ABI if you use r16 through r31 to pass floats to functions that aren't built with an awareness of these registers. So we simply don't! Easy choice, there.